### PR TITLE
[BugFix] Fix transaction stream load lock leak

### LINKED ISSUE #53155: [BUG] CN nodes can not release query_pool Fragment memory
### Steps to reproduce the behavior (Required)
1. Use Stream Load to write data until the disk is full
2. Drop table（clear disk）
3. recreate table
4. query_pool holds the timeout Fragment and does not release the memory
5. There is also a Prometheus exporter in my cluster that has been grabbing indicators and will query the tables under the information_schema database.

b7bd068bdf1336886d2c881294a59a2
f346015dfa0a9697fe97119796973bc

```log
I20241125 09:36:51.547662 140311655126592 plan_fragment_executor.cpp:373] cancel(): fragment_instance_id=aa4da8e4-1b28-8f97-e73d-0dd24b62d894
I20241125 09:36:51.547668 140311655126592 fragment_mgr.cpp:580] FragmentMgr cancel worker going to cancel timeout fragment aa4da8e4-1b28-8f97-e73d-0dd24b62d894
I20241125 09:36:51.547673 140311655126592 plan_fragment_executor.cpp:373] cancel(): fragment_instance_id=074d8d9d-b1d4-a6b9-1650-8a2ce1f416b4
I20241125 09:36:51.547678 140311655126592 fragment_mgr.cpp:580] FragmentMgr cancel worker going to cancel timeout fragment 074d8d9d-b1d4-a6b9-1650-8a2ce1f416b4
I20241125 09:36:51.547683 140311655126592 plan_fragment_executor.cpp:373] cancel(): fragment_instance_id=f442a4a2-f77b-7c13-a226-ce95f96a7f9b
I20241125 09:36:51.547698 140311655126592 fragment_mgr.cpp:580] FragmentMgr cancel worker going to cancel timeout fragment f442a4a2-f77b-7c13-a226-ce95f96a7f9b
I20241125 09:36:51.547706 140311655126592 plan_fragment_executor.cpp:373] cancel(): fragment_instance_id=d945dc8f-654b-df90-6028-93782edc47ae
I20241125 09:36:51.547712 140311655126592 fragment_mgr.cpp:580] FragmentMgr cancel worker going to cancel timeout fragment d945dc8f-654b-df90-6028-93782edc47ae
I20241125 09:36:51.547718 140311655126592 plan_fragment_executor.cpp:373] cancel(): fragment_instance_id=36434211-973c-02db-4324-b1ef5e67278d
I20241125 09:36:51.547724 140311655126592 fragment_mgr.cpp:580] FragmentMgr cancel worker going to cancel timeout fragment 36434211-973c-02db-4324-b1ef5e67278d
I20241125 09:36:51.547731 140311655126592 plan_fragment_executor.cpp:373] cancel(): fragment_instance_id=fa467e34-bfbf-94d3-b3d1-8afe18a97586
I20241125 09:36:51.547737 140311655126592 fragment_mgr.cpp:580] FragmentMgr cancel worker going to cancel timeout fragment fa467e34-bfbf-94d3-b3d1-8afe18a97586
I20241125 09:36:51.547744 140311655126592 plan_fragment_executor.cpp:373] cancel(): fragment_instance_id=db44962c-94f4-4d43-f232-b56dfbe4da9b
I20241125 09:36:51.547750 140311655126592 fragment_mgr.cpp:580] FragmentMgr cancel worker going to cancel timeout fragment db44962c-94f4-4d43-f232-b56dfbe4da9b
I20241125 09:36:51.547757 140311655126592 plan_fragment_executor.cpp:373] cancel(): fragment_instance_id=a6430d4d-78bb-23dc-6a67-34462a96bbb3
I20241125 09:36:51.547761 140311655126592 fragment_mgr.cpp:580] FragmentMgr cancel worker going to cancel timeout fragment a6430d4d-78bb-23dc-6a67-34462a96bbb3
I20241125 09:36:51.547768 140311655126592 plan_fragment_executor.cpp:373] cancel(): fragment_instance_id=db431cf3-86cb-7076-bca3-38a439644fb9
I20241125 09:36:51.547773 140311655126592 fragment_mgr.cpp:580] FragmentMgr cancel worker going to cancel timeout fragment db431cf3-86cb-7076-bca3-38a439644fb9
I20241125 09:36:51.547778 140311655126592 plan_fragment_executor.cpp:373] cancel(): fragment_instance_id=bb489cac-aada-38ce-428b-f256a43acf87
I20241125 09:36:51.547783 140311655126592 fragment_mgr.cpp:580] FragmentMgr cancel worker going to cancel timeout fragment bb489cac-aada-38ce-428b-f256a43acf87
I20241125 09:36:51.547789 140311655126592 plan_fragment_executor.cpp:373] cancel(): fragment_instance_id=0c4a80ca-e36a-1c39-fe9e-25d80f79dbb4
I20241125 09:36:51.547794 140311655126592 fragment_mgr.cpp:580] FragmentMgr cancel worker going to cancel timeout fragment 0c4a80ca-e36a-1c39-fe9e-25d80f79dbb4
I20241125 09:36:51.547801 140311655126592 plan_fragment_executor.cpp:373] cancel(): fragment_instance_id=f744af0e-f89e-f648-a3bf-722e0512bea8
I20241125 09:36:51.547807 140311655126592 fragment_mgr.cpp:580] FragmentMgr cancel worker going to cancel timeout fragment f744af0e-f89e-f648-a3bf-722e0512bea8
I20241125 09:36:51.547813 140311655126592 plan_fragment_executor.cpp:373] cancel(): fragment_instance_id=ec45fb1d-4c39-cc69-9e19-1afd7b9fc4a5
I20241125 09:36:51.547819 140311655126592 fragment_mgr.cpp:580] FragmentMgr cancel worker going to cancel timeout fragment ec45fb1d-4c39-cc69-9e19-1afd7b9fc4a5
I20241125 09:36:51.547826 140311655126592 plan_fragment_executor.cpp:373] cancel(): fragment_instance_id=264ccb9b-1249-0f3c-5a20-265f24e8a3ba
I20241125 09:36:51.547843 140311655126592 fragment_mgr.cpp:580] FragmentMgr cancel worker going to cancel timeout fragment 264ccb9b-1249-0f3c-5a20-265f24e8a3ba
I20241125 09:36:51.547850 140311655126592 plan_fragment_executor.cpp:373] cancel(): fragment_instance_id=1d413bea-78fb-673e-ec05-0bc20c2b21a1
I20241125 09:36:51.547856 140311655126592 fragment_mgr.cpp:580] FragmentMgr cancel worker going to cancel timeout fragment 1d413bea-78fb-673e-ec05-0bc20c2b21a1
I20241125 09:36:51.547862 140311655126592 plan_fragment_executor.cpp:373] cancel(): fragment_instance_id=d04614ff-f6a3-4686-90a5-3c6a4d717fae
I20241125 09:36:51.547867 140311655126592 fragment_mgr.cpp:580] FragmentMgr cancel worker going to cancel timeout fragment d04614ff-f6a3-4686-90a5-3c6a4d717fae
I20241125 09:36:51.547873 140311655126592 plan_fragment_executor.cpp:373] cancel(): fragment_instance_id=6e46f73b-cbb5-1840-48ae-c5baaffec2ab
I20241125 09:36:51.547878 140311655126592 fragment_mgr.cpp:580] FragmentMgr cancel worker going to cancel timeout fragment 6e46f73b-cbb5-1840-48ae-c5baaffec2ab
I20241125 09:36:51.547884 140311655126592 plan_fragment_executor.cpp:373] cancel(): fragment_instance_id=294429e4-38ff-5d96-f2dd-f156d0af6d99
I20241125 09:36:51.547889 140311655126592 fragment_mgr.cpp:580] FragmentMgr cancel worker going to cancel timeout fragment 294429e4-38ff-5d96-f2dd-f156d0af6d99
I20241125 09:36:51.547896 140311655126592 plan_fragment_executor.cpp:373] cancel(): fragment_instance_id=d542de6b-0253-9207-cb0d-b14012d454a3
I20241125 09:36:51.547901 140311655126592 fragment_mgr.cpp:580] FragmentMgr cancel worker going to cancel timeout fragment d542de6b-0253-9207-cb0d-b14012d454a3
I20241125 09:36:51.547907 140311655126592 plan_fragment_executor.cpp:373] cancel(): fragment_instance_id=724066be-eee3-ad13-aef3-27f768ff5cb9
I20241125 09:36:51.547912 140311655126592 fragment_mgr.cpp:580] FragmentMgr cancel worker going to cancel timeout fragment 724066be-eee3-ad13-aef3-27f768ff5cb9
I20241125 09:36:51.547917 140311655126592 plan_fragment_executor.cpp:373] cancel(): fragment_instance_id=a945c58b-fbd5-703f-def4-712bca7e9787
I20241125 09:36:51.547922 140311655126592 fragment_mgr.cpp:580] FragmentMgr cancel worker going to cancel timeout fragment a945c58b-fbd5-703f-def4-712bca7e9787
I20241125 09:36:51.547928 140311655126592 plan_fragment_executor.cpp:373] cancel(): fragment_instance_id=254c819d-de35-27a4-6584-053618239989
I20241125 09:36:51.547934 140311655126592 fragment_mgr.cpp:580] FragmentMgr cancel worker going to cancel timeout fragment 254c819d-de35-27a4-6584-053618239989
I20241125 09:36:51.547941 140311655126592 plan_fragment_executor.cpp:373] cancel(): fragment_instance_id=0747a6d5-8161-2007-e265-5f9548f22b98
I20241125 09:36:51.547947 140311655126592 fragment_mgr.cpp:580] FragmentMgr cancel worker going to cancel timeout fragment 0747a6d5-8161-2007-e265-5f9548f22b98
```

From the log analysis, cancel_worker has been working, but has not been able to clean up the Fragment's memory. Looking at the code from the log, it is confirmed that the cancel method in the plan_fragment_executor class is being executed. What I currently suspect is that `[result_mgr()->cancel(_runtime_state->fragment_instance_id())](https://github.com/StarRocks/starrocks/blob/3.3.2/be/src/runtime/plan_fragment_executor.cpp#L403)` has not completed the cleanup action.

I tried to capture the thread stack to take a look, but no exception was found, because cancel_worker sleeps for 1 second after each cleanup, and the captured thread stack is in sleep

### Expected behavior (Required)
query_pool release Fragment memory when timeout

### Real behavior (Required)
cancel worker always try to clean Fragment memory, but does not works well

### StarRocks version (Required)
 - 3.3.2
 - I have added a function built into the CN code in the current version, but this function is not currently used so it should not cause this problem.

### PR DESCRIPTION
**Pull request by banmoy**
## Why I'm doing:
In transaction stream load, a normal load request needs to go through the `on_header`, `on_chunk_data`, and `handle` processes of the `TransactionStreamLoadAction`.  To prevent concurrent load requests, a lock will be acquired in `on_header`, and released in `handle`. But for an abnormal load request, `handle` may be not called, and the lock will not be released. For example, the client crashes before transmitting all of data, only `on_header` and `on_chunk_data` are called, and `handle` will not be called. This can lead to resources not being cleaned up after the load timeout because the cleanup needs acquire the lock (see [TransactionMgr::_clean_stream_context](https://github.com/StarRocks/starrocks/blob/main/be/src/runtime/stream_load/transaction_mgr.cpp#L393)).

## What I'm doing:
Move the lock release to `TransactionStreamLoadAction::free_handler_ctx` which will be always called regardless of whether an exception occurs or not.

Fixes #53155

## What type of PR is this:

- [x] BugFix
- [ ] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [x] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [x] I have added test cases for my bug fix or my new feature
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function
- [ ] This is a backport pr

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto-backported to the target branch
  - [x] 3.4
  - [x] 3.3
  - [x] 3.2
  - [x] 3.1
  - [ ] 3.0

### CODE REVIEW

File: be/src/http/action/transaction_stream_load.cpp [hunk 1]
```diff
@@ -584,6 +565,23 @@ void TransactionStreamLoadAction::on_chunk_data(HttpRequest* req) {
     ctx->last_active_ts = MonotonicNanos();
     ctx->received_data_cost_nanos += ctx->last_active_ts - start_read_data_time;
     ctx->total_received_data_cost_nanos += ctx->last_active_ts - start_read_data_time;
+    VLOG(1) << "Receive http chunk, " << ctx->brief() << ", total expected bytes: " << ctx->body_bytes
+            << ", total received bytes: " << ctx->total_receive_bytes;
+}
+
+void TransactionStreamLoadAction::free_handler_ctx(void* param) {
+    StreamLoadContext* ctx = (StreamLoadContext*)param;
```
Comment (kevincai) [inline be/src/http/action/transaction_stream_load.cpp:573]: static_cast?

File: be/src/http/action/transaction_stream_load.cpp [hunk 2]
```diff
@@ -584,6 +565,23 @@ void TransactionStreamLoadAction::on_chunk_data(HttpRequest* req) {
...
+    if (ctx == nullptr) {
```
Comment (kevincai) [inline be/src/http/action/transaction_stream_load.cpp:574]: shall check nullptr against `param` before convert to StreamLoadContext?

**Review by kevincai — APPROVED** [review]

File: be/src/http/action/transaction_stream_load.cpp (hunk 1 above)
Comment (banmoy) [reply be/src/http/action/transaction_stream_load.cpp]: fixed

File: be/src/http/action/transaction_stream_load.cpp (hunk 2 above)
Comment (banmoy) [reply be/src/http/action/transaction_stream_load.cpp:574]: static_cast can be applied to nullptr so it's safe to check after the cast

**Review by wyb — APPROVED** [review]

**Comment by jaogoy** [comment]
ignore backport check: 3.1.16

**Pull request merged**